File: studio/AuraStudio/Sources/AuraStudio/Services/ImageResizer.swift
import Foundation
import ImageIO
import UniformTypeIdentifiers

/// Redimensiona fotos a una resolucion optima para el LCD de 320x240
/// del iPod usando ImageIO/CoreGraphics -- nativo de macOS, sin
/// depender de ffmpeg para algo que el sistema ya resuelve bien.
/// Preserva aspecto (no recorta ni deforma) y siempre convierte a JPEG,
/// que es uno de los dos formatos que el visor de Aura decodifica
/// (D-028 en el firmware; el otro es BMP, mucho mas pesado sin
/// compresion, no tiene sentido para fotos).
struct ImageResizer {
    enum ResizeError: Error, LocalizedError {
        case cannotReadImage
        case cannotEncodeOutput

        var errorDescription: String? {
            switch self {
            case .cannotReadImage: return "No se pudo leer la imagen de origen."
            case .cannotEncodeOutput: return "No se pudo generar el JPEG de salida."
            }
        }
    }

    /// 320x240 es la resolucion nativa del LCD -- default cuando no se
    /// pasa `maxDimension` explicito (p.ej. desde tests). Con la
    /// preferencia de calidad de foto (D-191/D-192), el llamador real
    /// (`LibraryViewModel.process`) pasa 320 o 640 segun lo que haya
    /// elegido el usuario; en ambos casos se preserva aspecto sin
    /// escalar hacia arriba fotos que ya sean chicas.
    static let maxDimension: CGFloat = 320

    static func resizeToLCDOptimal(sourceURL: URL, destinationURL: URL,
                                    maxDimension: CGFloat = maxDimension, quality: CGFloat = 0.85) throws {
        guard let source = CGImageSourceCreateWithURL(sourceURL as CFURL, nil) else {
            throw ResizeError.cannotReadImage
        }
        try write(source: source, destinationURL: destinationURL, maxDimension: maxDimension, quality: quality)
    }

    /// ST-033: misma conversion, desde bytes en memoria (poster de video
    /// descargado). Salida JPEG baseline con el lado mayor <= `maxDimension`
    /// (640 = maximo que admite el firmware, CONTRATO-firmware-studio.md).
    static func resizeToLCDOptimal(data: Data, destinationURL: URL,
                                    maxDimension: CGFloat = maxDimension, quality: CGFloat = 0.85) throws {
        guard let source = CGImageSourceCreateWithData(data as CFData, nil) else {
            throw ResizeError.cannotReadImage
        }
        try write(source: source, destinationURL: destinationURL, maxDimension: maxDimension, quality: quality)
    }

    private static func write(source: CGImageSource, destinationURL: URL,
                              maxDimension: CGFloat, quality: CGFloat) throws {

        let options: [CFString: Any] = [
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceThumbnailMaxPixelSize: maxDimension,
            kCGImageSourceCreateThumbnailWithTransform: true,
        ]

        guard let thumbnail = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary) else {
            throw ResizeError.cannotReadImage
        }

        try encodeBaselineJPEG(thumbnail, quality: quality).write(to: destinationURL)
    }

    // MARK: - Recorte cuadrado (contrato v18, ST-140)

    /// El JPEG cuadrado de lado `side`, recortado al centro desde el lado
    /// corto de la fuente (fill + center-crop, nunca estirado ni con
    /// bandas). Es la primitiva que usan la biblioteca local y el sync:
    /// carátulas de álbum (`cover.jpg` 320), fotos de artista (128) y la
    /// copia local de `.portadas/` (lado corto, tope 1000).
    ///
    /// Nunca escala hacia arriba: una fuente cuyo lado corto sea menor que
    /// `side` sale con ese lado corto. La orientación EXIF se respeta y se
    /// hornea en los píxeles, igual que en `resizeToLCDOptimal`.
    static func squareCrop(data: Data, side: Int, quality: CGFloat = 0.85) throws -> Data {
        guard let source = CGImageSourceCreateWithData(data as CFData, nil) else {
            throw ResizeError.cannotReadImage
        }
        return try squareCrop(source: source, side: side, quality: quality)
    }

    static func squareCrop(sourceURL: URL, side: Int, quality: CGFloat = 0.85) throws -> Data {
        guard let source = CGImageSourceCreateWithURL(sourceURL as CFURL, nil) else {
            throw ResizeError.cannotReadImage
        }
        return try squareCrop(source: source, side: side, quality: quality)
    }

    static func squareCrop(data: Data, destinationURL: URL, side: Int, quality: CGFloat = 0.85) throws {
        try squareCrop(data: data, side: side, quality: quality).write(to: destinationURL)
    }

    static func squareCrop(sourceURL: URL, destinationURL: URL, side: Int, quality: CGFloat = 0.85) throws {
        try squareCrop(sourceURL: sourceURL, side: side, quality: quality).write(to: destinationURL)
    }

    private static func squareCrop(source: CGImageSource, side: Int, quality: CGFloat) throws -> Data {
        guard let (width, height) = orientedPixelSize(of: source) else {
            throw ResizeError.cannotReadImage
        }

        let plan = SquareCropPlan(width: width, height: height, maxSide: side)
        guard !plan.isEmpty else { throw ResizeError.cannotReadImage }

        // ImageIO limita el lado MAYOR de la miniatura; lo que aca hay que
        // fijar es el lado CORTO (el que sobrevive al recorte). Se pide el
        // mayor proporcional, redondeado hacia arriba para que el corto
        // nunca quede por debajo de lo pedido y el recorte no tenga que
        // agrandar nada.
        let longSide = max(width, height), shortSide = min(width, height)
        let maxPixelSize = plan.needsResize
            ? Int((Double(longSide) * Double(plan.outputSide) / Double(shortSide)).rounded(.up))
            : longSide

        let options: [CFString: Any] = [
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceThumbnailMaxPixelSize: maxPixelSize,
            kCGImageSourceCreateThumbnailWithTransform: true,
        ]
        guard let thumbnail = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary) else {
            throw ResizeError.cannotReadImage
        }

        // El recorte se replantea sobre lo que ImageIO devolvio de verdad
        // (su redondeo puede diferir en un pixel del calculado arriba).
        let crop = SquareCropPlan(width: thumbnail.width, height: thumbnail.height, maxSide: plan.outputSide)
        guard !crop.isEmpty else { throw ResizeError.cannotReadImage }

        let square: CGImage
        if crop.needsCrop {
            guard let cropped = thumbnail.cropping(to: CGRect(x: crop.cropX, y: crop.cropY,
                                                              width: crop.cropSide, height: crop.cropSide)) else {
                throw ResizeError.cannotEncodeOutput
            }
            square = cropped
        } else {
            square = thumbnail
        }

        // El lado final es SIEMPRE el del plan sobre la fuente: el contrato
        // v18 fija medidas exactas (320x320, 128x128) y un pixel de menos
        // por el redondeo de la miniatura seria un incumplimiento.
        let resized = square.width == plan.outputSide ? square : try scale(square, toSide: plan.outputSide)
        return try encodeBaselineJPEG(resized, quality: quality)
    }

    /// Medidas de la imagen **ya orientadas** -- una foto vertical de
    /// camara viene guardada horizontal con la rotacion en EXIF, y el
    /// recorte tiene que planearse sobre lo que se ve.
    private static func orientedPixelSize(of source: CGImageSource) -> (width: Int, height: Int)? {
        guard let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
              let width = properties[kCGImagePropertyPixelWidth] as? Int,
              let height = properties[kCGImagePropertyPixelHeight] as? Int,
              width > 0, height > 0 else {
            return nil
        }
        // Orientaciones 5-8: la imagen guardada esta girada 90 grados, asi
        // que lo que se ve tiene los lados intercambiados.
        let orientation = properties[kCGImagePropertyOrientation] as? Int ?? 1
        return (5...8).contains(orientation) ? (height, width) : (width, height)
    }

    private static func scale(_ image: CGImage, toSide side: Int) throws -> CGImage {
        guard side > 0,
              let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
              let context = CGContext(data: nil, width: side, height: side,
                                       bitsPerComponent: 8, bytesPerRow: 0, space: colorSpace,
                                       bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else {
            throw ResizeError.cannotEncodeOutput
        }
        context.interpolationQuality = .high
        context.draw(image, in: CGRect(x: 0, y: 0, width: side, height: side))
        guard let scaled = context.makeImage() else { throw ResizeError.cannotEncodeOutput }
        return scaled
    }

    /// El JPEG de `image`, aplanado sobre blanco y garantizado baseline.
    private static func encodeBaselineJPEG(_ image: CGImage, quality: CGFloat) throws -> Data {
        // PLAN-sync-media-hardening.md PARTE 2A: una fuente PNG/GIF con
        // canal alfa (transparencia) llegaba tal cual al codificador de
        // JPEG -- que no tiene canal alfa, asi que el RGB debajo de los
        // pixeles transparentes queda a su criterio (con frecuencia
        // negro/indefinido en vez del fondo blanco esperado). Se aplana
        // sobre blanco ANTES de codificar, sin excepcion (para una
        // imagen ya opaca esto no cambia nada visible).
        let flattened = flattenOntoWhite(image) ?? image

        let output = NSMutableData()
        guard let destination = CGImageDestinationCreateWithData(output, UTType.jpeg.identifier as CFString, 1, nil) else {
            throw ResizeError.cannotEncodeOutput
        }

        // D-291 en Aura-Firmware (aura_photos.c:171-259): el visor solo
        // decodifica JPEG BASELINE -- un progresivo (marcador SOF2) sale
        // como "Formato no soportado". El codificador de ImageIO no
        // garantiza baseline por defecto para toda entrada; forzarlo
        // explicito (en vez de dejarlo a su criterio) es la unica forma
        // de asegurarlo siempre.
        let jfifProperties: [CFString: Any] = [kCGImagePropertyJFIFIsProgressive: false]
        let properties: [CFString: Any] = [
            kCGImageDestinationLossyCompressionQuality: quality,
            kCGImagePropertyJFIFDictionary: jfifProperties,
        ]
        CGImageDestinationAddImage(destination, flattened, properties as CFDictionary)

        guard CGImageDestinationFinalize(destination) else {
            throw ResizeError.cannotEncodeOutput
        }
        return output as Data
    }

    /// Compone `image` sobre un fondo blanco opaco, del mismo tamaño,
    /// descartando cualquier canal alfa. `nil` si CoreGraphics no pudo
    /// armar el contexto (fuente exotica) -- el llamador sigue de largo
    /// con la imagen original en vez de fallar el sync entero por esto.
    private static func flattenOntoWhite(_ image: CGImage) -> CGImage? {
        let width = image.width, height = image.height
        guard width > 0, height > 0,
              let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
              let context = CGContext(data: nil, width: width, height: height,
                                       bitsPerComponent: 8, bytesPerRow: 0, space: colorSpace,
                                       bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue) else {
            return nil
        }
        let rect = CGRect(x: 0, y: 0, width: width, height: height)
        context.setFillColor(CGColor(red: 1, green: 1, blue: 1, alpha: 1))
        context.fill(rect)
        context.draw(image, in: rect)
        return context.makeImage()
    }
}

File: studio/windows/AuraStudio.App/Platform/ImageResizer.cs
using AuraStudio.Core.Library;
using Windows.Graphics.Imaging;
using Windows.Storage.Streams;

namespace AuraStudio.App.Platform;

public sealed class ImageResizeException(string message) : Exception(message);

/// <summary>
/// Redimensiona fotos a la resolución del LCD del iPod usando WIC
/// (<c>Windows.Graphics.Imaging</c>) — nativo de Windows, sin depender de
/// ffmpeg para algo que el sistema ya resuelve bien. Port de
/// <c>ImageResizer.swift</c>, que allá usa ImageIO.
///
/// <para>La salida siempre es JPEG: es uno de los dos formatos que decodifica
/// el visor de Aura (D-028 del firmware; el otro es BMP, sin compresión, que
/// para fotos no tiene sentido).</para>
/// </summary>
public static class ImageResizer
{
    /// <summary>La misma calidad que macOS, para que las dos apps produzcan lo mismo.</summary>
    public const double DefaultQuality = 0.85;

    public static Task ResizeToLcdOptimalAsync(
        string sourcePath, string destinationPath,
        int maxDimension = ImageResizePlan.DefaultMaxDimension,
        double quality = DefaultQuality)
        => ResizeAsync(() => File.ReadAllBytes(sourcePath), destinationPath, maxDimension, quality);

    /// <summary>
    /// ST-033: la misma conversión desde bytes en memoria (el póster de un
    /// video descargado, que nunca toca el disco como original).
    /// </summary>
    public static Task ResizeToLcdOptimalAsync(
        byte[] source, string destinationPath,
        int maxDimension = ImageResizePlan.DefaultMaxDimension,
        double quality = DefaultQuality)
        => ResizeAsync(() => source, destinationPath, maxDimension, quality);

    private static async Task ResizeAsync(Func<byte[]> readSource, string destinationPath,
                                          int maxDimension, double quality)
    {
        byte[] jpeg = await EncodeAsync(readSource(), maxDimension, quality).ConfigureAwait(false);

        string? directory = Path.GetDirectoryName(destinationPath);
        if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
        await File.WriteAllBytesAsync(destinationPath, jpeg).ConfigureAwait(false);
    }

    /// <summary>El JPEG resultante, ya verificado como baseline.</summary>
    public static async Task<byte[]> EncodeAsync(byte[] source, int maxDimension, double quality)
    {
        if (source.Length == 0) throw new ImageResizeException("No se pudo leer la imagen de origen.");

        using var input = new InMemoryRandomAccessStream();
        await input.WriteAsync(ToBuffer(source));
        input.Seek(0);

        BitmapDecoder decoder;
        try
        {
            decoder = await BitmapDecoder.CreateAsync(input);
        }
        catch (Exception ex)
        {
            // WIC a veces no trae mensaje; una frase con dos puntos colgando
            // en pantalla es peor que la frase sola.
            string detail = string.IsNullOrWhiteSpace(ex.Message) ? "" : $" ({ex.Message.Trim()})";
            throw new ImageResizeException($"No se pudo leer la imagen de origen.{detail}");
        }

        // El tamaño destino se calcula sobre las medidas YA orientadas (una foto
        // vertical de cámara viene guardada horizontal con la orientación en
        // EXIF). El factor de escala es el mismo en ambos espacios —la
        // orientación solo intercambia o espeja los lados—, así que se aplica
        // sobre las medidas crudas, que es lo que espera BitmapTransform.
        (int orientedWidth, int orientedHeight) = ImageResizePlan.TargetSize(
            (int)decoder.OrientedPixelWidth, (int)decoder.OrientedPixelHeight, maxDimension);
        if (orientedWidth == 0 || orientedHeight == 0)
            throw new ImageResizeException("La imagen de origen no tiene un tamaño válido.");

        double scale = (double)orientedWidth / decoder.OrientedPixelWidth;
        var transform = new BitmapTransform
        {
            ScaledWidth = (uint)Math.Max(1, (int)Math.Round(decoder.PixelWidth * scale)),
            ScaledHeight = (uint)Math.Max(1, (int)Math.Round(decoder.PixelHeight * scale)),
            InterpolationMode = BitmapInterpolationMode.Fant
        };

        using SoftwareBitmap scaled = await decoder.GetSoftwareBitmapAsync(
            BitmapPixelFormat.Bgra8, BitmapAlphaMode.Straight, transform,
            ExifOrientationMode.RespectExifOrientation, ColorManagementMode.ColorManageToSRgb);

        return await EncodeJpegAsync(scaled, quality);
    }

    // --- Recorte cuadrado (contrato v18, ST-140) ---

    /// <summary>
    /// El JPEG cuadrado de lado <paramref name="side"/>, recortado al centro
    /// desde el lado corto de la fuente (fill + center-crop, nunca estirado ni
    /// con bandas). Es la primitiva que usan la biblioteca local y el sync:
    /// carátulas de álbum (<c>cover.jpg</c> 320), fotos de artista (128) y la
    /// copia local de <c>.portadas\</c> (lado corto, tope 1000).
    ///
    /// <para>Nunca escala hacia arriba: una fuente cuyo lado corto sea menor
    /// que <paramref name="side"/> sale con ese lado corto. La orientación EXIF
    /// se respeta y se hornea en los píxeles. Port de
    /// <c>ImageResizer.squareCrop(data:side:quality:)</c> en macOS — mismos
    /// números, misma aritmética (<see cref="SquareCropPlan"/>).</para>
    /// </summary>
    public static async Task<byte[]> EncodeSquareAsync(byte[] source, int side, double quality)
    {
        if (source.Length == 0) throw new ImageResizeException("No se pudo leer la imagen de origen.");

        using var input = new InMemoryRandomAccessStream();
        await input.WriteAsync(ToBuffer(source));
        input.Seek(0);

        BitmapDecoder decoder;
        try
        {
            decoder = await BitmapDecoder.CreateAsync(input);
        }
        catch (Exception ex)
        {
            string detail = string.IsNullOrWhiteSpace(ex.Message) ? "" : $" ({ex.Message.Trim()})";
            throw new ImageResizeException($"No se pudo leer la imagen de origen.{detail}");
        }

        // El plan se calcula sobre las medidas YA orientadas: lo que se recorta
        // es lo que se ve, no cómo está guardado el archivo.
        var plan = SquareCropPlan.For((int)decoder.OrientedPixelWidth, (int)decoder.OrientedPixelHeight, side);
        if (plan.IsEmpty)
            throw new ImageResizeException("La imagen de origen no tiene un tamaño válido.");

        // Lo que hay que fijar es el lado CORTO (el que sobrevive al recorte),
        // no el mayor: se lleva EXACTO al lado pedido y el largo se redondea
        // hacia arriba, para que el recorte nunca tenga que agrandar nada y el
        // resultado mida exactamente lo que el contrato v18 fija (320, 128).
        // Las medidas van en el espacio CRUDO, que es lo que espera
        // BitmapTransform; la orientación solo intercambia o espeja los lados,
        // así que el lado corto es el mismo en los dos espacios.
        int rawWidth = (int)decoder.PixelWidth, rawHeight = (int)decoder.PixelHeight;
        int scaledWidth, scaledHeight;
        if (rawWidth <= rawHeight)
        {
            scaledWidth = plan.OutputSide;
            scaledHeight = Math.Max(plan.OutputSide,
                                    (int)Math.Ceiling((double)rawHeight * plan.OutputSide / rawWidth));
        }
        else
        {
            scaledHeight = plan.OutputSide;
            scaledWidth = Math.Max(plan.OutputSide,
                                   (int)Math.Ceiling((double)rawWidth * plan.OutputSide / rawHeight));
        }

        // Bounds recorta sobre la imagen YA escalada. Un cuadrado centrado es
        // el mismo antes y después de la orientación EXIF (girar 90° o espejar
        // no mueve el centro), así que da igual en qué espacio lo aplique WIC:
        // lo único que puede cambiar de lado es el píxel sobrante de un margen
        // impar, que es medio píxel de diferencia con macOS y no se ve.
        var crop = SquareCropPlan.For(scaledWidth, scaledHeight, plan.OutputSide);
        var transform = new BitmapTransform
        {
            ScaledWidth = (uint)scaledWidth,
            ScaledHeight = (uint)scaledHeight,
            InterpolationMode = BitmapInterpolationMode.Fant,
            Bounds = new BitmapBounds
            {
                X = (uint)crop.CropX,
                Y = (uint)crop.CropY,
                Width = (uint)crop.CropSide,
                Height = (uint)crop.CropSide
            }
        };

        using SoftwareBitmap square = await decoder.GetSoftwareBitmapAsync(
            BitmapPixelFormat.Bgra8, BitmapAlphaMode.Straight, transform,
            ExifOrientationMode.RespectExifOrientation, ColorManagementMode.ColorManageToSRgb);

        return await EncodeJpegAsync(square, quality);
    }

    /// <summary>
    /// Escribe el JPEG cuadrado directo a un archivo, creando su carpeta.
    /// </summary>
    public static async Task SquareCropAsync(byte[] source, string destinationPath, int side,
                                             double quality = DefaultQuality)
    {
        byte[] jpeg = await EncodeSquareAsync(source, side, quality).ConfigureAwait(false);

        string? directory = Path.GetDirectoryName(destinationPath);
        if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
        await File.WriteAllBytesAsync(destinationPath, jpeg).ConfigureAwait(false);
    }

    public static Task SquareCropAsync(string sourcePath, string destinationPath, int side,
                                       double quality = DefaultQuality)
        => SquareCropAsync(File.ReadAllBytes(sourcePath), destinationPath, side, quality);

    /// <summary>El JPEG de la imagen, aplanada sobre blanco y ya verificada como baseline.</summary>
    private static async Task<byte[]> EncodeJpegAsync(SoftwareBitmap bitmap, double quality)
    {
        using SoftwareBitmap flattened = FlattenOntoWhite(bitmap);

        using var output = new InMemoryRandomAccessStream();
        var options = new BitmapPropertySet
        {
            { "ImageQuality", new BitmapTypedValue(quality, Windows.Foundation.PropertyType.Single) }
        };
        BitmapEncoder encoder = await BitmapEncoder.CreateAsync(BitmapEncoder.JpegEncoderId, output, options);
        encoder.SetSoftwareBitmap(flattened);
        await encoder.FlushAsync();

        byte[] jpeg = await ToBytesAsync(output);

        // D-291: el visor del firmware solo decodifica JPEG baseline. macOS se
        // lo pide explícitamente a ImageIO; el codificador de WIC no expone esa
        // opción, así que la garantía se consigue verificando la salida. Si
        // algún día dejara de ser baseline, se sabe acá y no en el iPod.
        if (!JpegMarkers.IsBaseline(jpeg))
            throw new ImageResizeException(
                "El JPEG generado no es baseline y el iPod no podría mostrarlo.");

        return jpeg;
    }

    /// <summary>
    /// Compone la imagen sobre blanco opaco y descarta el canal alfa.
    ///
    /// <para>Una fuente PNG/GIF con transparencia llegaba tal cual al
    /// codificador JPEG —que no tiene canal alfa—, así que el RGB debajo de los
    /// píxeles transparentes quedaba a su criterio (con frecuencia negro, en
    /// vez del fondo blanco esperado). Se aplana <b>siempre</b>: para una imagen
    /// ya opaca no cambia nada visible.</para>
    /// </summary>
    private static SoftwareBitmap FlattenOntoWhite(SoftwareBitmap source)
    {
        var buffer = new Windows.Storage.Streams.Buffer(
            (uint)(source.PixelWidth * source.PixelHeight * 4));
        source.CopyToBuffer(buffer);
        byte[] pixels = ToBytes(buffer);

        for (int i = 0; i + 3 < pixels.Length; i += 4)
        {
            byte alpha = pixels[i + 3];
            if (alpha == 255) continue;

            // BGRA, alfa no premultiplicado: c sobre blanco = c·a + 255·(1-a).
            for (int channel = 0; channel < 3; channel++)
                pixels[i + channel] = (byte)((pixels[i + channel] * alpha + 255 * (255 - alpha) + 127) / 255);
            pixels[i + 3] = 255;
        }

        var flattened = new SoftwareBitmap(
            BitmapPixelFormat.Bgra8, source.PixelWidth, source.PixelHeight, BitmapAlphaMode.Ignore);
        flattened.CopyFromBuffer(ToBuffer(pixels));
        return flattened;
    }

    private static IBuffer ToBuffer(byte[] bytes)
    {
        var writer = new DataWriter();
        writer.WriteBytes(bytes);
        return writer.DetachBuffer();
    }

    private static byte[] ToBytes(IBuffer buffer)
    {
        byte[] bytes = new byte[buffer.Length];
        DataReader.FromBuffer(buffer).ReadBytes(bytes);
        return bytes;
    }

    private static async Task<byte[]> ToBytesAsync(IRandomAccessStream stream)
    {
        stream.Seek(0);
        var buffer = new Windows.Storage.Streams.Buffer((uint)stream.Size);
        await stream.ReadAsync(buffer, (uint)stream.Size, InputStreamOptions.None);
        return ToBytes(buffer);
    }
}
